File: src/viz/runtime/evaluator.ts
import type { Binding } from "../../bindings";
import { resolveEase, type EaseFn } from "./ease";

const clamp = (v: number, lo: number, hi: number): number =>
  v < lo ? lo : v > hi ? hi : v;
const clamp01 = (v: number): number => clamp(v, 0, 1);

/**
 * Evaluates one binding's transform each frame, holding the attack/release
 * smoothing envelope as internal state. Construct one per binding; call
 * `evaluate(signal)` with the source band's current 0..1 level.
 */
export class BindingEvaluator {
  private smoothed = 0;
  private readonly ease: EaseFn;

  constructor(
    private readonly binding: Binding,
    ease: EaseFn = resolveEase(binding.transform.ease)
  ) {
    this.ease = ease;
  }

  /**
   * Map a raw 0..1 signal to the target's output value. Pipeline: clamp to
   * [0,1] → attack/release envelope (when configured) → `^exponent` → ease →
   * lerp into [outMin, outMax] → clamp to that range.
   */
  evaluate(signal: number): number {
    const t = this.binding.transform;
    let s = clamp01(signal);

    if (t.smoothing) {
      const coef = s > this.smoothed ? t.smoothing.attack : t.smoothing.release;
      this.smoothed += (s - this.smoothed) * coef;
      s = this.smoothed;
    }

    if (t.exponent !== undefined) s = Math.pow(s, t.exponent);
    s = this.ease(s);

    const out = t.outMin + (t.outMax - t.outMin) * s;
    return clamp(out, Math.min(t.outMin, t.outMax), Math.max(t.outMin, t.outMax));
  }
}

File: tools/blender/addon/evaluator.py
from .ease import resolve_ease


def _clamp(v, lo, hi):
    return lo if v < lo else (hi if v > hi else v)


def _clamp01(v):
    # NaN-safe: mirrors evaluator.ts clamp01 (v > 0 ? (v < 1 ? v : 1) : 0).
    # A non-finite signal (e.g. an empty bucket's 0/0) maps to 0.
    if v > 0:
        return v if v < 1 else 1
    return 0


class BindingEvaluator:
    """Mirror of src/viz/runtime/evaluator.ts. One instance per binding; holds the
    attack/release envelope as state. Call evaluate(signal) with the 0..1 level."""

    def __init__(self, binding, ease=None):
        self.binding = binding
        self.smoothed = 0
        transform = binding["transform"]
        self.ease = ease if ease is not None else resolve_ease(transform.get("ease"))

    def evaluate(self, signal):
        t = self.binding["transform"]
        s = _clamp01(signal)

        # Gate-then-rescale before the envelope, mirroring evaluator.ts: below
        # the threshold is exactly 0; the surviving range remaps to 0..1.
        gate = t.get("gate")
        if gate is not None:
            s = max(0, s - gate) / (1 - gate)

        smoothing = t.get("smoothing")
        if smoothing is not None:
            coef = smoothing["attack"] if s > self.smoothed else smoothing["release"]
            self.smoothed += (s - self.smoothed) * coef
            s = self.smoothed

        exponent = t.get("exponent")
        if exponent is not None:
            s = s ** exponent

        s = self.ease(s)

        out = t["outMin"] + (t["outMax"] - t["outMin"]) * s
        return _clamp(out, min(t["outMin"], t["outMax"]), max(t["outMin"], t["outMax"]))
